### PR TITLE
add a method to collect the DNS names from a certificate

### LINKED ISSUE #2: Add a means to collect SANs from a certificate
We (Linkerd) currently depend on the changes in https://github.com/briansmith/webpki/pull/91 to read a client certificate's DNS SANs from a server-terminated TLS connection. This PR was superseded by @Geal's https://github.com/briansmith/webpki/pull/103, but this also was never merged. I'd love to find a path forward for these changes.

### PR DESCRIPTION
**Pull request by hawkw**
This branch adds an `EndEntityCert::dns_names` method, which returns a
list of the DNS names provided in the subject alternative names
extension of the certificate.

This branch is based on work done by @seanmonstar in
briansmith/webpki#91 and @Geal in briansmith/webpki#103. I've simply
updated this change to track the `main` branch of the `rustls/webpki`
repository.

Closes #2 

Authored-by: Geoffroy Couprie <geo.couprie@gmail.com>
Co-authored-by: Sean McArthur <sean@seanmonstar.com>
Co-authored-by: Eliza Weisman <eliza@buoyant.io>
Signed-off-by: Eliza Weisman <eliza@buoyant.io>

### CODE REVIEW

File: src/name/dns_name.rs [hunk 1]
```diff
@@ -147,6 +147,131 @@ impl<'a> From<DnsNameRef<'a>> for &'a str {
...
+/// a wildcard.
+///
+/// A `WildcardDnsName` is guaranteed to be syntactically valid. The validity rules are
+/// specified in [RFC 5280 Section 7.2], except that underscores are also
+/// allowed, and following [RFC 6125].
+///
+/// `WildcardDnsName` stores a copy of the input it was constructed from in a `String`
+/// and so it is only available when the `alloc` default feature is enabled.
+///
+/// `Eq`, `PartialEq`, etc. are not implemented because name comparison
```
Comment (samlh) [inline src/name/dns_name.rs:181]: Drive-by: This comment doesn't match the derive below?

File: src/name/verify.rs [hunk 2]
```diff
@@ -272,6 +277,33 @@ fn iterate_names(
     }
 }
 
+#[cfg(feature = "alloc")]
+pub fn list_cert_dns_names<'names>(
+    cert: &'names crate::EndEntityCert<'names>,
+) -> Result<Vec<GeneralDnsNameRef<'names>>, Error> {
```
Comment (djc) [inline src/name/verify.rs:283]: How hard would it be to rewire this to return an `impl Iterator` instead?

**Comment by cpu** [comment]
@hawkw :wave: gentle poke for this branch. Do you think you'll have a chance to address the earlier feedback? It looks like a useful change :+1:

File: src/name/verify.rs (hunk 2 above)
Comment (hawkw) [reply src/name/verify.rs:283]: it certainly could be rewritten to return `impl Iterator` --- I wanted to start by coping the original diff from upstream exactly, but would be happy to revise.

**Comment by hawkw** [comment]
> @hawkw wave gentle poke for this branch. Do you think you'll have a chance to address the earlier feedback? It looks like a useful change +1

sure, i'm happy to pick this back up!

**Comment by cpu** [comment]
I'd like to see this work land so I've adopted the branch to try and address the last bits of feedback. If any of the original authors want to pick this back up again I'm happy to step back to a reviewer role. In the meantime I made a second PR because I couldn't push commits to the linkerd fork for this branch. PTAL at https://github.com/rustls/webpki/pull/42

I'm going to close this PR to consolidate discussion on the new one but as mentioned, also happy to come back to this branch if Eliza or someone else has cycles to iterate. Thanks for getting it so close to being done!

**Comment by cpu** [comment]
Just wanted to mention in case folks are watching this PR - we landed the feature in #42.
